explainer.md

```diff
@@ -48,218 +49,298 @@ Provide web apps with efficient access to built-in (software and hardware) media
 
-An **AudioTrackReader** converts a MediaStreamTrack into a ReadableStream of AudioPacket.
+-   An **AudioEncoder** encodes AudioPackets to produce EncodedAudioChunks.
 
-An **AudioEncoder** is a TransformStream from AudioPacket to EncodedAudioChunk.
+-   A **VideoEncoder** encodes VideoFrames to produce EncodedVideoChunks.
 
-An **AudioDecoder** is a TransformStream from EncodedAudioChunk to AudioPacket.
+-   An **AudioDecoder** decodes EncodedAudioChunks to produce AudioPackets.
 
-An **AudioTrackWriter** converts a WritableStream of AudioPacket into a MediaStreamTrack.
+-   A **VideoDecoder** decodes EncodedVideoChunks to produce VideoFrames.
 
-A **VideoTrackReader** converts a MediaStreamTrack into a ReadableStream of VideoFrame.
 
-A **VideoEncoder** is a TransformStream from VideoFrame to EncodedVideoChunk.
+WebCodecs will also define mechanisms for importing content from getUserMedia().
 
-A **VideoDecoder** is a TransformStream from EncodedVideoChunk to VideoFrame.
+-   An **AudioTrackReader** converts an audio MediaStreamTrack into a ReadableStream of AudioPacket.
 
-A **VideoTrackWriter** converts a WritableStream of VideoFrame into a MediaStreamTrack.
+-   A **VideoTrackReader** converts a video MediaStreamTrack into a ReadableStream of VideoFrame.
 
 ## Examples
-### Example of video rendering to Canvas for low-latency live streaming or cloud gaming
+### Example of video rendering to Canvas for extremely low-latency streaming (e.g. cloud gaming)
 
 ```javascript
-// App defines a simple WritableSink that renders new frames as soon as they arrive.
-class CanvasRendererSink {
-  constructor(canvas) {
-    this._context = canvas.getContext('bitmaprenderer');
-  }
+// The app provides error handling.
+function onDecoderError(error) { ... }
 
-  write(videoFrame) {
-    _context.transferFromImageBitmap(videoFrame.image);
-    return Promise.resolve();
-  }
-}
+// App provides stream of encoded chunks to decoder.
+function streamEncodedChunks(decodeCallback) { ... }
 
-// The app provides ReadableStreams of encoded video
-// in the form of byte arrays defined by a codec such as vp8
-// (not in a media container such as mp4 or webm).
-const {encodedVideo} = ...;
-
-// App sets up canvas on the page for use by the rendering sink.
-const canvas = document.getElementById("canvas");
+// The document contains a canvas for displaying VideoFrames.
+const canvasElement = document.getElementById("canvas");
+const canvasContext = canvasElement.getContext('bitmaprenderer');
 
-// Wrap the rendering sink in a writable stream so it can be connected to the graph.
-const renderingStream = new WritableStream(new CanvasRendererSink(canvas));
+function paintFrameToCanvas(videoFrame) {
+  // Paint every video frame ASAP for lowest latency.
+  canvasContext.transferFromImageBitmap(videoFrame.image);
```

Comment:
Nit: this will probably be takeAsImageBitmap(), returning a Promise.

Reply:
videoFrame.transferToImageBitmap(), rather than guaranteeing that an ImageBitmap is always available synchronously. We may prefer getImageBitmap() to imply a copy, I'm not sure about that aspect at this point.

Reply:
Ah, of course. Done (transferToImageBitmap() for now)